### ISSUE TITLE
"Connection refused" from all sockets on Docker host

### ISSUE THREAD

**Issue by pcgeek86** [docker/docker#20611]
On my Macbook Pro running Mac OS X El Capitan, with a Docker Host running inside VirtualBox, I created a MySQL and a WordPress container. When I try to access WordPress, from my browser, or the `curl` command line, I get "connection refused."

As another test, I ran a Redis cache container, and get the same results (see below).

# Docker Version Output
```
Trevors-MBP:~ trevorsullivan$ docker version
Client:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:14:21 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 22:37:33 2016
 OS/Arch:      linux/amd64
```

# Docker Info Output

```
Trevors-MBP:~ trevorsullivan$ docker info
Containers: 10
 Running: 2
 Paused: 0
 Stopped: 8
Images: 4
Server Version: 1.10.2
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 85
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 4.1.18-boot2docker
Operating System: Boot2Docker 1.10.2 (TCL 6.4.1); master : 611be10 - Mon Feb 22 22:47:06 UTC 2016
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 996.1 MiB
Name: dockerhost
ID: OYEZ:5CVO:SWXW:O46K:EHBM:SIEL:GEE5:ZA4N:MCSF:DI4Y:FWJ5:SYU3
Debug mode (server): true
 File Descriptors: 24
 Goroutines: 38
 System Time: 2016-02-23T15:55:09.102357116Z
 EventsListeners: 0
 Init SHA1:
 Init Path: /usr/local/bin/docker
 Docker Root Dir: /mnt/sda1/var/lib/docker
Labels:
 provider=virtualbox
```

# Redis Cache Repro

Here's my attempt to launch and connect to a Redis instance, running inside a Docker container.

```
Trevors-MBP:~ trevorsullivan$ docker run -d redis
66c86171dc22385871f755e385cf4445a97760f65204b93604f00460ce828a72
Trevors-MBP:~ trevorsullivan$ docker-machine ip dockerhost
192.168.99.100
Trevors-MBP:~ trevorsullivan$ redis-cli -h 192.168.99.100
Could not connect to Redis at 192.168.99.100:6379: Connection refused
```

<img width="885" alt="screen shot 2016-02-23 at 8 38 56 am" src="https://cloud.githubusercontent.com/assets/466713/13257223/332dd5de-da0b-11e5-9e32-27a95b4922f8.png">

Any thoughts on how I can resolve this networking problem, so I can get an application running properly?

Cheers,
Trevor Sullivan
Microsoft MVP: PowerShell
https://trevorsullivan.net
https://twitter.com/pcgeek86

**Comment by thaJeztah** [comment]
Looks like you forgot to *publish* the container's ports, so even though the container is listening for connections, its ports are not published, so not accessible from outside the virtual machine.

For example, this will start an nginx container, and publishes it's port '80' on port '8081' on the host;

    docker run -d -p 8081:80 nginx

You can then connect to that container through http://192.189.99.100:8081 (assuming that's the IP address of your machine/VM)

**Comment by thaJeztah** [comment]
For example, see https://docs.docker.com/engine/userguide/containers/usingdocker/#running-a-web-application-in-docker, and

**Comment by pcgeek86** [comment]
You're spot on, but last night, I tried that with WordPress, to no avail.

Here's Redis working:

```
Trevors-MBP:~ trevorsullivan$ docker run -d -p 6379:6379 redis
f64bb8762e5ae67ab0c5f9c82acca27d01a371373faa726c85de32447c35994e
Trevors-MBP:~ trevorsullivan$ redis-cli -h 192.168.99.100
192.168.99.100:6379>
```

I'm going to give WordPress another shot.

**Comment by thaJeztah** [comment]
Let me close this issue for now, but happy to reopen if it's still an issue

**Comment by abhinavkorpal** [comment]
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 11:02:43 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 11:02:43 2017
 OS/Arch:      linux/amd64
 Experimental: false


Containers: 7
 Running: 1
 Paused: 0
 Stopped: 6
Images: 7
Server Version: 17.03.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 51
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 977c511eda0925a723debdc94d09459af49d082a
runc version: a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-66-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.713 GiB
Name: bsetec-Lenovo-G50-80
ID: WGQ4:XMID:KUKC:PSVU:AA4J:GHJ6:CDYH:IHWV:RPG7:YUZI:WSFN:TLYW
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false


curl -v http://192.168.99.100:8080/ping
*   Trying 192.168.99.100...
* connect to 192.168.99.100 port 8080 failed: Connection refused
* Failed to connect to 192.168.99.100 port 8080: Connection refused
* Closing connection 0
curl: (7) Failed to connect to 192.168.99.100 port 8080: Connection refused

**Comment by Crabime** [comment]
@abhinavkorpal have you resolved your question? here i met the same question, already published VM 8080 to container using `docker run -td -p 8080:8080 --name tomcat imageName`, but still refused connect when using `curl -v 192.168.99.100 8080`

**Comment by thaJeztah** [comment]
@Crabime make sure that the service that is running in the container is listening on `0.0.0.0` not on `127.0.0.1` (as that would only make it accessible from inside the container)

**Comment by aloglua** [comment]
Hi;
In my case port configuration is "-p  8080:8080 -p 4848:4848"  (payara micro)
When i run netstat  -a -n in docker quick start terminal for windows 7, I can not see  the ports 8080 and 4848 in the list. ( I tried Docker Tooling in Eclipse the result is same)
Does that mean the container does not listen those ports? If so what should I do?
Docker version 18.03.0 CE
Thanks a lot.

**Comment by nlgl** [comment]
Mr. thaJeztah
In my case  ports configuration is "-p 8080:8080 - p 4848:4848". I tried eclipse docker tooling and result is same. I even disabled firewall.
When I run netstats -a -n in docker quickstart  terminal on window 7, I do not see these port in the list.
Does that mean the container does not listen these ports. If so what should i do? 
Docker version is 18.03.0-ce  OS/Arch falsews/amd64

Thank you.

I got here via google search "Failed to connect to 192.168.99.100 port" search. Hope I am on the right place to ask question.
